filebeat/crawler/prospector_log.go

```diff
@@ -125,192 +125,72 @@ func (p *ProspectorLog) scan() {
 	} else {
-		p.resumeHarvesting(h, 0)
-	}
-}
-
-// checkOldFile returns true if the given file is currently not harvested
-// and the last time was modified before ignore_older
-func (p *ProspectorLog) checkOldFile(h *harvester.Harvester) bool {
-
-	// Resuming never needed if ignore_older disabled
-	if p.config.IgnoreOlderDuration == 0 {
-		return false
-	}
-
-	modTime := h.Stat.Fileinfo.ModTime()
-
-	// Make sure modification time is before the last scan started to not pick it up twice
-	if !modTime.Before(p.lastscan) {
-		return false
-	}
-
-	// Only should be checked if older then ignore_older
-	if time.Since(modTime) <= p.config.IgnoreOlderDuration {
-		return false
+		logp.Debug("prospector", "Ignore file because ignore_older reached: %s", state.Source)
 	}
-
-	return true
 }
 
-// checkExistingFile checks if a harvester has to be started for a already known file
-// For existing files the following options exist:
-// * Last reading position is 0, no harvester has to be started as old harvester probably still busy
-// * The old known modification time is older then the current one. Start at last known position
-// * The new file is not the same as the old file, means file was renamed
-// ** New file is actually really a new file, start a new harvester
-// ** Renamed file has a state, continue there
-func (p *ProspectorLog) checkExistingFile(h *harvester.Harvester, newFile *input.File, oldFile *input.File) {
-
-	logp.Debug("prospector", "Update existing file for harvesting: %s", h.Path)
-
-	// We assume it is the same file, but it wasn't
-	if !oldFile.IsSameFile(newFile) {
-
-		logp.Debug("prospector", "File previously found: %s", h.Path)
+// harvestExistingFile continues harvesting a file with a known state if needed
+func (p *ProspectorLog) harvestExistingFile(newState input.FileState, oldState input.FileState) {
 
-		if previousFile, err := p.getPreviousFile(h.Path, h.Stat.Fileinfo); err == nil {
-			p.continueExistingFile(h, previousFile)
-		} else {
-			// File is not the same file we saw previously, it must have rotated and is a new file
-			logp.Debug("prospector", "Launching harvester on rotated file: %s", h.Path)
-
-			// Forget about the previous harvester and let it continue on the old file - so start a new channel to use with the new harvester
-			h.Stat.Ignore()
-
-			// Start a new harvester on the path
-			h.Start()
-		}
+	logp.Debug("prospector", "Update existing file for harvesting: %s, offset: %v", newState.Source, oldState.Offset)
 
-		// Keep the old file in missingFiles so we don't rescan it if it was renamed and we've not yet reached the new filename
-		// We only need to keep it for the remainder of this iteration then we can assume it was deleted and forget about it
-		p.missingFiles[h.Path] = oldFile.FileInfo
-
-	} else if h.Stat.Finished() && oldFile.FileInfo.ModTime() != h.Stat.Fileinfo.ModTime() {
+	// No harvester is running for the file, start a new harvester
+	if oldState.Finished {
```

Comment:
oldState.Finished set from another go-routine? If so, race detector should complain.